server/src/main/java/org/candlepin/util/Transactional.java

```diff
@@ -0,0 +1,323 @@
+     *  this transactional wrapper
+     */
+    public Transactional<O> onComplete(Listener listener) {
+        this.onCommit(listener);
+        this.onRollback(listener);
+
+        return this;
+    }
+
+    /**
+     * Sets this wrapper to automatically commit the transaction if an uncaught exception occurs.
+     * By default, the wrapper will rollback the transaction on exception, but this method can be
+     * used to specify the behavior.
+     *
+     * @return
+     *  this transactional wrapper
+     */
+    public Transactional<O> commitOnException() {
+        this.commitOnException = true;
+        return this;
+    }
+
+    /**
+     * Sets this wrapper to allow use of a pre-existing active transaction. When reuse mode is
+     * enabled, a warning will be logged if a transactional block is executed with an active
+     * transaction already present.
+     *
+     * @return
+     *  this transactional wrapper
+     */
+    public Transactional<O> allowExistingTransactions() {
+        this.exclusive = false;
+        return this;
+    }
+
+    /**
+     * Commits the transaction and notifies listeners
+     */
+    private void commitTransaction(EntityTransaction transaction) {
+        transaction.commit();
+
+        for (Listener listener : this.commitListeners) {
+            listener.transactionComplete(Status.STATUS_COMMITTED);
+        }
+    }
+
+    /**
+     * Rolls back the transaction and notifies listeners
+     */
+    private void rollbackTransaction(EntityTransaction transaction) {
+        transaction.rollback();
+
+        for (Listener listener : this.rollbackListeners) {
+            listener.transactionComplete(Status.STATUS_ROLLEDBACK);
+        }
+    }
+
+
+    /**
+     * Executes this transactional, passing the specified arguments through to the action.
+     *
+     * @param args
+     *  the args to pass through to the action
+     *
+     * @return
+     *  the output of the action
+     */
+    public O execute(Object... args) {
+        if (this.action == null) {
+            throw new IllegalStateException("no action provided");
+        }
+
+        O output = null;
+        EntityTransaction transaction = this.entityManager.getTransaction();
+
+        if (!transaction.isActive()) {
+            transaction.begin();
+        }
+        else {
+            String errmsg = "Transactional block executed with a transaction already started";
+            if (this.exclusive) {
+                throw new IllegalStateException(errmsg);
+            }
+
+            log.warn(errmsg);
+        }
+
+        try {
+            output = this.action.execute(args);
+        }
+        catch (Exception e) {
+            if (!this.commitOnException) {
+                transaction.setRollbackOnly();
+            }
+
+            throw e;
+        }
+        finally {
+            if (!transaction.isActive()) {
+                String errmsg = "Transactional block completed without an active transaction";
```

Comment:
It would be good to add a test case that covers this block (e.g. when an action itself has already commited or rolled back the current transaction), just to be safe.

Reply:
Added